### ISSUE TITLE
Unknown macro variables give "unexpected token" error, not an error that they are unknown

### ISSUE THREAD

**Issue by huonw** [rust-lang/rust#18839]
```rust
#![feature(macro_rules)]

macro_rules! foo { () => { $e }}

fn main() {
    foo!()
}
```

```
unknown-macro-variable.rs:3:29: 3:30 error: unexpected token: `$`
unknown-macro-variable.rs:3 macro_rules! foo { () => { $e }}
                                                        ^
```

This is very confusing and does not provide much help to work out what is wrong. (Looks like this was introduced by 38ce6d9eac5d0bcfa0c102bc64393a987b4a43e3.)
